### ISSUE TITLE
OAuth doesn't work

### ISSUE THREAD

**Issue by frec-hansen** [blakmatrix/node-zendesk#393]
**Describe the Bug**
Keeps getting `Request processing failed: Zendesk Error (401): Not Authorized`

**Code I tried**
```javascript
// Not working
zendeskClient = createClient({
  endpointUri: "https://<subdomain>.zendesk.com",
  token: "xxx",
  oauth: true,
});

// Not working
zendeskClient = createClient({
  endpointUri: "https://<subdomain>.zendesk.com/api/v2",
  token: "xxx",
  oauth: true,
});

// Not working
zendeskClient = createClient({
  subdomain: "<subdomain>",
  token: "xxx",
  oauth: true,
});
```

**Expected Behavior**
Expected it to work

**Actual Behavior**
Got `Request processing failed: Zendesk Error (401): Not Authorized`

**Environment Information**
- `node-zendesk` version: 5.0.5
- Node.js version: v18.16.1
- Operating System: MacOS
- Any other relevant software versions?

**Additional Context**
Also the token is the token you get from here and create client and `secret` field right?
<img width="291" alt="Screenshot 2023-11-16 at 6 28 20 PM" src="https://github.com/blakmatrix/node-zendesk/assets/97467751/87763c5f-7d96-4d3c-9642-37dbda1a8c3d">

**Comment by blakmatrix** [comment]
Hello @frec-hansen ,

Thank you for reaching out with your issue regarding the `401: Not Authorized` error in the `node-zendesk` library. After reviewing the details, it seems the problem stems from the way OAuth tokens are being handled, i.e. not a bug--you need to provision an OAuth token.

To help you navigate this, I've updated the library and added a detailed example script. Here are the key updates and some important guidelines:

1. **New OAuthClients Class**: I've introduced a new `OAuthClients` class in the library. This class offers functionalities for managing OAuth clients, including listing, creating, updating, and deleting OAuth clients, and generating client secrets.

2. **Detailed Example Script**: An example script has been updated with comprehensive JSDoc comments. It demonstrates how to create a read-only OAuth token, which involves retrieving your OAuth client's numeric ID using its unique identifier. You can find this script @ https://github.com/blakmatrix/node-zendesk/blob/master/examples/create-oauth-token.js.

3. **Provisioning an OAuth Token**: It's important to note that for using the library, you need to provision an OAuth token specifically for this purpose. The secret token you mentioned is used in a _different context_ with a back and forth OAuth Client. The example script will guide you through creating an OAuth token with the necessary scopes. Initially, you'll need your username/password or username/API token (and appropriat admin access of course) to generate an OAuth token in this method.

4. **Setting Appropriate Permissions**: The example demonstrates creating a token with read-only access for users and tickets. Depending on your needs, you can adjust these scopes to grant more permissions. Please refer to the Zendesk documentation on [Creating and using OAuth tokens with the API](https://developer.zendesk.com/documentation/ticketing/working-with-oauth/creating-and-using-oauth-tokens-with-the-api/) for more detailed guidance or alternitive methods for provisioning(i.e. curl).

5. **Using the OAuth Token for Authentication**: Once you've generated the OAuth token, you can use it to authenticate in the client. Keep in mind the permissions (scopes) assigned to the token will dictate what actions you can perform through the API.

I hope these updates and instructions will help you resolve the authorization issues. Please pull the latest version of the `node-zendesk` library to access these new features and improvements. If you have any further questions or face additional issues, feel free to reach out.

Best regards,

Farrin

**Comment by frec-hansen** [comment]
Thanks for the detailed reply and updated example. I will revert back and use the username/password auth since it is a lot easier. I just want minimum setup to use it. BTW. I want to point out that some of the documents probably also need to be updated. When I was using [users.createOrUpdate](https://blakmatrix.github.io/node-zendesk/code/classes/clients_core_users.Users.html#createorupdate) according to doc like this:
```javascript
zendeskClient.users.createOrUpdate({ name:"test", email:"example@email.com" });
```
I got a 400 response.
Turns out the attributes needs to be wrapped in user object:
```javascript
zendeskClient.users.createOrUpdate({ user: { name:"test", email:"example@email.com" }  });
```
